x-pack/platform/plugins/shared/lens/public/editor_frame_service/editor_frame/config_panel/use_initialize_chart.ts

```diff
@@ -0,0 +1,133 @@
+
+type LensAttributes = TypedLensSerializedState['attributes'];
+
+export interface InitializeChartLogicArgs {
+  /**
+   * Indicates if the query is in text-based language (ESQL).
+   */
+  isTextBasedLanguage: boolean;
+  /**
+   * The query to be executed.
+   */
+  query: AggregateQuery | Query;
+  /**
+   * Attributes for the ESQL data grid, if applicable.
+   */
+  dataGridAttrs: ESQLDataGridAttrs | undefined;
+  /**
+   * Indicates if the dataGridAttrs havw been initialized.
+   */
+  isInitialized: boolean;
+  /**
+   * Current attributes of the chart.
+   */
+  currentAttributes: LensAttributes | undefined;
+  /**
+   * Reference to the previous query.
+   */
+  prevQueryRef: MutableRefObject<AggregateQuery | Query>;
+  /**
+   * Function to set errors that occur during initialization.
+   */
+  setErrors: (errors: Error[]) => void;
+  /**
+   * Function to set the initialization state.
+   */
+  setIsInitialized: (isInitialized: boolean) => void;
+  /**
+   * Function to run the query and update the chart.
+   */
+  runQuery: (
+    q: AggregateQuery,
+    abortController?: AbortController,
+    shouldUpdateAttrs?: boolean
+  ) => Promise<void>;
+}
+
+/**
+ * Encapsulates the logic for initializing the chart/data grid based on ESQL query.
+ *
+ */
+export const createInitializeChartFunction = ({
+  isTextBasedLanguage,
+  query,
+  dataGridAttrs,
+  isInitialized,
+  currentAttributes,
+  runQuery,
+  prevQueryRef,
+  setErrors,
+  setIsInitialized,
+}: InitializeChartLogicArgs) => {
+  return async (abortController?: AbortController) => {
+    if (isInitialized) {
+      // If already initialized, do nothing
+      return;
+    }
+    if (isTextBasedLanguage && isOfAggregateQueryType(query) && !dataGridAttrs) {
+      try {
+        const shouldUpdateAttrs = Boolean(currentAttributes?.state.needsRefresh);
+        await runQuery(query, abortController, shouldUpdateAttrs);
+      } catch (e) {
+        setErrors([e]);
+      }
+      prevQueryRef.current = query;
+    }
+    setIsInitialized(true);
+  };
+};
+
+export function useInitializeChart({
+  isTextBasedLanguage,
+  query,
+  dataGridAttrs,
+  isInitialized,
+  currentAttributes,
+  runQuery,
+  prevQueryRef,
+  setErrors,
+  setIsInitialized,
+}: InitializeChartLogicArgs) {
+  const initializeChartFunc = useCallback(
+    (abortController: AbortController) => {
+      const func = createInitializeChartFunction({
+        isTextBasedLanguage,
+        query,
+        dataGridAttrs,
+        isInitialized,
+        currentAttributes,
+        runQuery,
+        prevQueryRef,
+        setErrors,
+        setIsInitialized,
+      });
+      func(abortController);
+    },
+    [
+      isTextBasedLanguage,
+      query,
+      dataGridAttrs,
+      isInitialized,
+      currentAttributes,
+      runQuery,
+      prevQueryRef,
+      setErrors,
+      setIsInitialized,
+    ]
+  );
+  useEffect(() => {
+    const abortController = new AbortController();
```

Comment:
nit: this can be moved inside the `initializeChartFunc` content? I do not see any other context where it will be passed elsewhere.